File: 백준/Silver/1874. 스택 수열/스택 수열.py
N = int(input())
target = []   # 우선, 오름차순으로 1~N 만큼 넣을 스택.
ans = []   # "+" or "-" 값을 넣을 스택.
flag = 0
t = 1  # 1부터 오름차순으로 넣으니까.
for _ in range(N):
    n = int(input())
    while t <= n :  # 오름차순 push 니까 1부터 push, 이 때 입력한 수 만날 때 까지만 push.
        target.append(t)     #
        ans.append("+")
        t += 1

    if target[-1] == n:  # 원하던 숫자 pop
        target.pop()
        ans.append("-")

    else:
        print("NO")    # 마지막이 원하던 숫자보다 크다면 원하던 건 그 밑에 있어서 완성될 수 없다.
        flag = 1
        break

if flag == 0:
    for i in ans:
        print(i)

File: 백준/Silver/1874. 스택 수열/스택 수열.java
import java.util.Scanner;
import java.util.Stack;

public class Main{
    public static void main(String[] args) {
        // 입력값.
        Scanner sc  = new Scanner(System.in);
        int n  = sc.nextInt();
        int A[]  = new int[n];
        for(int i=0; i<n; i++){
            A[i] = sc.nextInt();
        }
        Stack<Integer> stack = new Stack<>();
        int num = 1; //start 1부터 하나씩 넣음.
        boolean result = true; // No이면 false flag 표시.

        StringBuffer bf = new StringBuffer(); // String 배열 넣을 것임.
        for(int i = 0; i<A.length; i++){
            int su = A[i];  // 입력 숫자 꺼냄.
            // if 숫자크기 만큼 스택에 넣고 해당 숫자를 꺼냄.
            if(su >= num){
                while (su>=num){
                   stack.push(num++);
                    bf.append("+\n");
                }
                stack.pop();
                bf.append("-\n");
            }
            // 스택 맨 위의 값(ju)이 입력 값보다 크면 NO/ 같으면 POP
             else {
                int ju = stack.pop();
                if(ju > su){
                    System.out.println("NO");
                    result = false;
                    break;
                } else {
                    bf.append("-\n");
                }
            }
        }
        //flag가 true면 bf 배열 출력.
        if(result) System.out.println(bf);
    }
}
